pkg/controller/instance/instance_controller.go

```diff
@@ -543,19 +516,112 @@ func isUpgradePlan(planName string) bool {
+// (ENABLE_WEBHOOKS=true) we have TWO WAYS of deciding that:
+//
+// 1. WITHOUT the webhook, we have the old logic which iterates over all existing PlanStatuses, searching for
+//    the first one with Status == InProgress. It might happen that multiple plans are in progress (e.g. cleanup and deploy)
+// 2. WITH the webhook, current plan lives in the Spec.PlanExecution.PlanName field
+func getPlanInProgress(i *v1beta1.Instance) *kudov1beta1.PlanStatus {
+	if areWebhooksEnabled() {
+		return i.PlanStatus(i.Spec.PlanExecution.PlanName)
+	}
+	return i.GetPlanInProgress()
+}
+
+// newExecutionPlan method returns a new execution plan (if exists) or nil otherwise. As long as we don't enforce webhooks
+// (ENABLE_WEBHOOKS=true) we have TWO WAYS of deciding which plan has to executed next:
+//
+// 1. WITHOUT the webhook, we have the old logic which tries to infer the change (parameter update, new OperatorVersion etc.)
+//    by diffing current state with the one from the snapshot.
+// 2. WITH the webhook, instance admission webhook has already decided on the plan and the result is in the
+//    Spec.PlanExecution.PlanName field.
+func newExecutionPlan(i *v1beta1.Instance, ov *v1beta1.OperatorVersion) (plan *string, err error) {
+	if areWebhooksEnabled() {
+		if plan, err = fetchNewExecutionPlan(i, ov); plan != nil && err == nil {
+			log.Printf("InstanceController: Fetched new execution plan '%s' from the spec for instance %s/%s", kudo.StringValue(plan), i.Namespace, i.Name)
+		}
+	} else {
+		if plan, err = inferNewExecutionPlan(i, ov); plan != nil && err == nil {
+			log.Printf("InstanceController: Inferred new execution plan '%s' from instance %s/%s state", kudo.StringValue(plan), i.Namespace, i.Name)
+		}
+	}
+
+	return
+}
+
+// fetchNewExecutionPlan method fetches a new execution plan from Instance.Spec.PlanExecution field. A plan is actually
+// new if the PlanExecution.UID has changed, otherwise it's just an old plan in progress.
+// - "planName", when there is a new plan that needs to be executed
+// - <nil>, no new plan found e.g. a plan is already in progress
+func fetchNewExecutionPlan(i *v1beta1.Instance, ov *v1beta1.OperatorVersion) (*string, error) {
+	snapshot, err := i.SnapshotSpec()
+	if err != nil {
+		return nil, &v1beta1.InstanceError{Err: fmt.Errorf("failed to unmarshal instance snapshot %s/%s: %v", i.Namespace, i.Name, err), EventName: kudo.String("UnexpectedState")}
+	}
+
+	// Instance deletion is an edge case where the admission webhook *can not* populate the Spec.PlanExecution.PlanName
+	// with the 'cleanup' plan. So we have to do it here ourselves. Only if:
+	// 1. Instance is being deleted
+	// 2. Cleanup plan exists in the operator version and has *never run* before
+	// 3. Cleanup hasn't been scheduled yet (first time the deletion is being reconciled)
+	// we set the Spec.PlanExecution.PlanName = 'cleanup'
+	hasToScheduleCleanupAfterDeletion := func() bool {
```

Comment:
Is there a specific reason that these are inline funcs? I'd rather have them outside as actual funcs or just plain code in this method.

Reply:
Since they're only used once inside this method, they would otherwise "pollute" the controller namespace. I prefer to inline them to give a "speaking name" to otherwise complicated check.

Reply:
Yeah, I get the "pollution" argument, although I blame go for it ;)

Still, I wouldn't mind having just the simple booleans inside this func instead of full functions that are only called once, but it's not a blocker for me. Just looks weird.